src/tensors/tensor.jl

```diff
@@ -350,52 +350,57 @@ function TensorMap(data::DenseArray, codom::ProductSpace{S,N₁}, dom::ProductSp
          size(data) == (dims(codom)..., dims(dom)...))
         throw(DimensionMismatch())
     end
+
     if sectortype(S) === Trivial
         data2 = reshape(data, (d1, d2))
         A = typeof(data2)
         return TensorMap{S,N₁,N₂,Trivial,A,Nothing,Nothing}(data2, codom, dom)
-    else
-        t = TensorMap(zeros, eltype(data), codom, dom)
-        ta = convert(Array, t)
-        l = length(ta)
-        dimt = dim(t)
-        basis = zeros(eltype(ta), (l, dimt))
-        qdims = zeros(real(eltype(ta)), (dimt,))
-        i = 1
-        for (c, b) in blocks(t)
-            for k in 1:length(b)
-                b[k] = 1
-                copy!(view(basis, :, i), reshape(convert(Array, t), (l,)))
-                qdims[i] = dim(c)
-                b[k] = 0
-                i += 1
-            end
-        end
-        rhs = reshape(data, (l,))
-        if FusionStyle(sectortype(t)) isa UniqueFusion
-            lhs = basis' * rhs
-        else
-            lhs = Diagonal(qdims) \ (basis' * rhs)
-        end
-        if norm(basis * lhs - rhs) > tol
-            throw(ArgumentError("Data has non-zero elements at incompatible positions"))
-        end
-        if eltype(lhs) != scalartype(t)
-            t2 = TensorMap(zeros, promote_type(eltype(lhs), scalartype(t)), codom, dom)
-        else
-            t2 = t
-        end
-        i = 1
-        for (c, b) in blocks(t2)
-            for k in 1:length(b)
-                b[k] = lhs[i]
-                i += 1
-            end
+    end
+
+    t = TensorMap(zeros, eltype(data), codom, dom)
```

Comment:
Maybe it is useful to wrap this in a function on its own, that can be used by the ad rules, and only keep the final `isapprox` check in the actual `TensorMap` constructor. No strong feelings about this one though.